lib/core/mqtt.cpp

```diff
@@ -0,0 +1,174 @@
+}
+
+mqtt::message_ptr Mqtt::messageToMessagePtr(const MqttMessage &message)
+{
+  rapidjson::Document payload;
+  payload.SetObject();
+  rapidjson::Value header;
+  header.SetObject();
+  header.AddMember("priority", message.header.priority, payload.GetAllocator());
+  header.AddMember("timestamp", message.header.timestamp, payload.GetAllocator());
+  payload.AddMember("header", header, payload.GetAllocator());
+  rapidjson::Value payload_json;
+  payload_json.CopyFrom(*message.payload, payload.GetAllocator());
+  payload.AddMember("payload", payload_json, payload.GetAllocator());
+  rapidjson::StringBuffer buffer;
+  auto writer = rapidjson::Writer<rapidjson::StringBuffer>(buffer);
+  payload.Accept(writer);
+  return mqtt::make_message(mqtt_topic_to_string[message.topic], buffer.GetString());
+}
+
+std::optional<MqttMessage> Mqtt::messagePtrToMessage(std::shared_ptr<const mqtt::message> message)
+{
+  const auto topic            = message->get_topic();
+  const auto mqtt_topic       = mqtt_string_to_topic[topic];
+  const auto message_contents = message->get_payload().c_str();
+  rapidjson::Document message_contents_json;
+  message_contents_json.Parse(message_contents);
+  if (message_contents_json.HasParseError()) {
+    logger_.log(core::LogLevel::kFatal, "Failed to parse MQTT message: parse error");
+    return std::nullopt;
+  }
+  if (!message_contents_json.HasMember("header")) {
+    logger_.log(core::LogLevel::kFatal, "Failed to parse MQTT message: missing header");
+    return std::nullopt;
+  }
+  const auto header = message_contents_json["header"].GetObject();  // error here
+  if (!header.HasMember("timestamp")) {
+    logger_.log(core::LogLevel::kFatal, "Failed to parse MQTT message: missing timestamp");
+  }
+  if (!header.HasMember("priority")) {
+    logger_.log(core::LogLevel::kFatal, "Failed to parse MQTT message: missing priority");
+  }
+  const auto timestamp = header["timestamp"].GetUint64();
+  const auto priority  = header["priority"].GetUint();
+  if (!message_contents_json.HasMember("payload")) {
+    logger_.log(core::LogLevel::kFatal, "Failed to parse MQTT message: missing payload");
+  }
+  if (priority > static_cast<std::uint8_t>(MqttMessagePriority::kLow)) {
+    logger_.log(core::LogLevel::kFatal, "Failed to parse MQTT message: invalid priority");
+    return std::nullopt;
+  }
+  MqttMessagePriority mqtt_priority = static_cast<MqttMessagePriority>(priority);
+  std::shared_ptr mqtt_payload      = std::make_shared<rapidjson::Document>();
+  mqtt_payload->CopyFrom(message_contents_json["payload"], mqtt_payload->GetAllocator());
+  MqttMessage::Header mqtt_header{timestamp, mqtt_priority};
+  return MqttMessage{mqtt_topic, mqtt_header, mqtt_payload};
+}
+
+MqttCallback::MqttCallback(ILogger &logger) : logger_(logger)
+{
+}
+
+void MqttCallback::connection_lost(const std::string &cause)
+{
+  if (!cause.empty()) {
+    logger_.log(core::LogLevel::kFatal, "Connection lost with MQTT broker: %s", cause.c_str());
```

Comment:
is there any way we can relay this back to the main thread? would be good to know when we lose connection, but that will also get caught by the heartbeat so might not be neccessary

Reply:
also, as this is in a different thread I should take a look at logger and figure out if it's threadsafe (it's not (my problem not yours (if you need work for the hackathon maybe???)))